nvflare/fuel/utils/config.py

```diff
@@ -0,0 +1,142 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Dict, List, Optional
+
+
+class ConfigFormat(Enum):
+    JSON = "JSON"
+    PYHOCON = "PYHOCON"
+    OMEGACONF = "OMEGACONF"
+
+    @classmethod
+    def _config_format_extensions(cls):
+        return {
+            "JSON": [".json", ".json.default"],
+            "PYHOCON": [".conf", ".conf.default"],
+            "OMEGACONF": [".yml", ".yml.default"],
+        }
+
+    @classmethod
+    def search_order(cls):
+        return [cls.JSON, cls.PYHOCON, cls.OMEGACONF]
+
+    @classmethod
+    def ordered_search_extensions(cls) -> List:
+        search_sequence: List[ConfigFormat] = cls.search_order()
+        extensions = []
+        for fmt in search_sequence:
+            exts = cls.get_extensions(fmt.name)
+            if exts:
+                for ext in exts:
+                    extensions.append((fmt, ext))
+
+        return extensions
+
+    @classmethod
+    def get_extensions(cls, fmt: str) -> Optional[List[str]]:
+        return cls._config_format_extensions().get(fmt, None)
+
+    @classmethod
+    def config_exts(cls) -> str:
+        search_sequence: List[ConfigFormat] = cls.search_order()
+        extensions = []
+        for fmt in search_sequence:
+            exts = cls.get_extensions(fmt.name)
+            if exts:
+                extensions.extend(exts)
+
+        return "|".join(iter(extensions))
+
+
+class Config(ABC):
+    @abstractmethod
+    def get_conf(self):
+        # Return the original underline config object representation if you prefer to use it directly
+        # Pyhocon → ConfigTree
+        # JSON → Dict
+        # OMEGACONF → Conf
+        pass
+
+    @abstractmethod
+    def get_format(self):
+        pass
+
+    @abstractmethod
+    def get_location(self) -> Optional[str]:
+        # return the real file path of the reel configuration.
+        # For example, if initial config_fed_json.json was initially input file path. but the real file path is
+        # config_fed_json.conf. the get location will be return config_fed_json.conf
+        pass
+
+    @abstractmethod
+    def to_dict(self) -> Dict:
+        pass
+
+    @abstractmethod
+    def to_conf(self, element: Dict) -> str:
+        pass
+
+    @abstractmethod
+    def get_int(self, key: str, default=None) -> Optional[int]:
+        pass
+
+    @abstractmethod
+    def get_float(self, key: str, conf=None, default=None) -> Optional[float]:
+        pass
+
+    @abstractmethod
+    def get_bool(self, key: str, default=None) -> Optional[bool]:
+        pass
+
+    @abstractmethod
+    def get_str(self, key: str, default=None) -> Optional[str]:
+        pass
+
+    @abstractmethod
+    def get_list(self, key: str, default=None) -> Optional[List]:
+        pass
+
+    @abstractmethod
+    def get_config(self, key: str, default=None):
+        pass
```

Comment:
What is the difference "get_config" VS "get_conf" VS "get_location"?


Say we have a "config_fed_server.json"

get_config will return "config_fed_server.json"?
get_conf will return a dictionary
get_location will return "config_fed_server.json"?

so get_config and get_location is the same thing?